ext/psycopg2: Implement BaseInstrumentor interface

cnnradams opened the pull request:
1. Implemented BaseInstrumentor interface to enable auto-instrumentation
2. Added integration tests (same tests as other db integrations)

Since the psycopg2 connection object can't have properties added to it (for example, ```self._db_api_integration = db_api_integration``` throws an error), I instead closed these properties around the tracing wrappers.

Resolves #633

ocelotl commented:
Hello @cnnradams 

What error do you get when you

> 1. Implemented BaseInstrumentor interface to enable auto-instrumentation
> 2. Added integration tests (same tests as other db integrations)
> 
> Since the psycopg2 connection object can't have properties added to it (for example, `self._db_api_integration = db_api_integration` throws an error), I instead closed these properties around the tracing wrappers.
> 
> Resolves #633

Hey, @cnnradams, what is the error you mention above?

cnnradams commented:
> Hello @cnnradams
> 
> What error do you get when you
> 
> > 1. Implemented BaseInstrumentor interface to enable auto-instrumentation
> > 2. Added integration tests (same tests as other db integrations)
> > 
> > Since the psycopg2 connection object can't have properties added to it (for example, `self._db_api_integration = db_api_integration` throws an error), I instead closed these properties around the tracing wrappers.
> > Resolves #633
> 
> Hey, @cnnradams, what is the error you mention above?

```
postgres/test_psycopg_functional.py:44: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../.tox/docker-tests/lib/python3.6/site-packages/wrapt/wrappers.py:567: in __call__
    args, kwargs)
../../opentelemetry-ext-dbapi/src/opentelemetry/ext/dbapi/__init__.py:118: in _wrap_connect
    return db_integration.wrapped_connection(wrapped, args, kwargs)
../../opentelemetry-ext-dbapi/src/opentelemetry/ext/dbapi/__init__.py:224: in wrapped_connection
    return TracedConnectionProxy(connection, self)
../../opentelemetry-ext-dbapi/src/opentelemetry/ext/dbapi/__init__.py:268: in __init__
    self._db_api_integration = db_api_integration
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <TracedConnectionProxy at 0x7faae3621208 for connection at 0x7faae3928660>
name = '_db_api_integration'
value = <opentelemetry.ext.dbapi.DatabaseApiIntegration object at 0x7faae3647898>

    def __setattr__(self, name, value):
        if name.startswith('_self_'):
            object.__setattr__(self, name, value)
    
        elif name == '__wrapped__':
            object.__setattr__(self, name, value)
            try:
                object.__delattr__(self, '__qualname__')
            except AttributeError:
                pass
            try:
                object.__setattr__(self, '__qualname__', value.__qualname__)
            except AttributeError:
                pass
    
        elif name == '__qualname__':
            setattr(self.__wrapped__, name, value)
            object.__setattr__(self, name, value)
    
        elif hasattr(type(self), name):
            object.__setattr__(self, name, value)
    
        else:
>           setattr(self.__wrapped__, name, value)
E           AttributeError: 'psycopg2.extensions.connection' object has no attribute '_db_api_integration'

../../../.tox/docker-tests/lib/python3.6/site-packages/wrapt/wrappers.py:190: AttributeError
```
basically, you can't seem to assign new properties to objects that were originally defined in C (same issue happens with sqlite3, does not happen with mysql since its connection was defined in python)

lzchen commented on ext/opentelemetry-ext-dbapi/src/opentelemetry/ext/dbapi/__init__.py:
```diff
@@ -221,7 +221,7 @@ def wrapped_connection(
         connection = connect_method(*args, **kwargs)
         self.get_connection_attributes(connection)
-        return TracedConnectionProxy(connection, self)
+        return get_traced_connection_proxy(connection, self)
```
Any reason for these wrapper functions?

cnnradams replied:
yes - see my above comment. tl;dr: Can't assign new properties to the psycopg2 connection object

lzchen commented on ext/opentelemetry-ext-grpc/src/opentelemetry/ext/grpc/grpcext/_interceptor.py:
```diff
@@ -164,6 +164,7 @@ def invoker(request_iterator, metadata):
         )
 
 
+# pylint: disable=abstract-method
```
Was this causing lint issues? Shouldn't really be included as part of this PR.

cnnradams replied:
It was, fixed by #696 - will remove

lzchen commented on ext/opentelemetry-ext-psycopg2/tests/test_psycopg2_integration.py:
```diff
@@ -12,17 +12,105 @@
-            trace_integration()
-            cnx = psycopg2.connect(database="test")
-            self.assertIsNotNone(cnx.cursor_factory)
+class TestPostgresqlIntegration(TestBase):
```
Good job using the TestBase!
